### PR TITLE
Only render portal children after DOM element is inserted

### PR DESCRIPTION
**Pull request by wmadden**
As documented here https://reactjs.org/docs/portals.html, the portal element is inserted in the DOM tree after the `<BodyPortal>`'s children are mounted, meaning that they will be mounted on a detached DOM node. For children that manage focus (like the `<ClearableInput>`) we must ensure they're only rendered after their host element is inserted in the DOM tree.

### CODE REVIEW

File: src/client/components/body-portal/body-portal.tsx [hunk 1]
```diff
@@ -32,7 +32,11 @@ export interface BodyPortalProps {
   isAboveAll?: boolean;
 }
 
-export class BodyPortal extends React.Component<BodyPortalProps, {}> {
+interface BodyPortalState {
+  isMounted: boolean;
```
Comment (adrianmroz) [inline src/client/components/body-portal/body-portal.tsx]: I would call that field `isAttached` and set it right after `document.body.appendChild`. 

I don't like keeping such state, but even docs say to do this ...

File: src/client/components/body-portal/body-portal.tsx [hunk 2]
```diff
@@ -58,14 +65,17 @@ export class BodyPortal extends React.Component<BodyPortalProps, {}> {
...
   componentWillUnmount() {
     document.body.removeChild(this.target);
     if (BodyPortal.aboveAll === this) BodyPortal.aboveAll = undefined;
   }
 
   render() {
+    const { isMounted } = this.state;
     Object.assign(this.target.style, normalizeStyles(this.props));
-    return ReactDOM.createPortal(this.props.children, this.target);
+    return ReactDOM.createPortal(isMounted && this.props.children, this.target);
```
Comment (adrianmroz) [inline src/client/components/body-portal/body-portal.tsx]: Is calling `createPortal` with `false` safe?

File: src/client/components/body-portal/body-portal.tsx [hunk 3]
```diff
@@ -44,6 +48,9 @@ export class BodyPortal extends React.Component<BodyPortalProps, {}> {
     super(props);
     this.target = document.createElement("div");
     this.target.className = classNames("body-portal", { "full-size": props.fullSize });
+    this.state = {
```
Comment (adrianmroz) [inline src/client/components/body-portal/body-portal.tsx]: I would go with `state` class property initialisation to be consistent. I know we have constructor for shenanigans with `target`, but keep it just for that.

File: src/client/components/body-portal/body-portal.tsx (hunk 2 above)
Comment (wmadden) [reply src/client/components/body-portal/body-portal.tsx]: I believe so. It has no effect in the running app (i.e. no warnings printed to console) and it's typed as a `ReactNode` which is a union including `boolean` (`type ReactNode = ReactChild | ReactFragment | ReactPortal | string | number | boolean | null | undefined;`).

The React API documentation says basically nothing for `createPortal()`, but [`React.Component`'s `render()`](https://reactjs.org/docs/react-component.html#render) method allows returning a boolean specifically to support the `test && <Child />` pattern.

File: src/client/components/body-portal/body-portal.tsx (hunk 3 above)
Comment (wmadden) [reply src/client/components/body-portal/body-portal.tsx]: 👍

File: src/client/components/body-portal/body-portal.tsx (hunk 1 above)
Comment (wmadden) [reply src/client/components/body-portal/body-portal.tsx]: 👍

**Comment by wmadden** [comment]
Addressed your comments @adrianmroz

**Pull request merged**